mesonbuild/msubprojects.py

```diff
@@ -131,6 +131,7 @@ def update_git(wrap, repo_dir, options):
     # is needed for http case were new remote branches wouldn't be discovered
     # otherwise. After this command, FETCH_HEAD is the revision we want.
     try:
+        git_output(['remote', 'update', 'origin'], repo_dir)
         git_output(['fetch', 'origin', revision], repo_dir)
```

Comment:
Shouldn't we 'just' call `git fetch origin` here?

Reply:
> Fetch only the revision we need, this avoids fetching useless branches

So now we fetch utterly everything, we can just remove the following line...

`git remote update origin` is identical to invoking `git fetch origin`, but previously (a long time ago) `git remote update` was the only way to fetch multiple remotes.

Reply:
Ok, changed to `git fetch origin`. I think we still need both calls however, consider this situation:

I already have 1.18 branch locally but it is outdated, `git fetch origin` will update `origin/1.18` but `git reset --hard 1.18` will reset to the local version. And I cannot do `git reset --hard origin/1.18` because that fails if 1.18 is a tag and not a branch. This is why I need `git fetch origin 1.18` to set FETCH_HEAD to the right commit, so `git reset --hard FETCH_HEAD` will always work regardless if revision is a tag, branch or commit id.

I think I also need `git fetch origin <revision>` in the case revision is a commit id not part of any references, don't think `git fetch origin` pulls that commit in that case, does it? But I admit this is corner case.

Reply:
This is all already the case before your commit. The question is why you need to fetch the entire remote repo before fetching the specific revision you're interested in. `git fetch origin 1.18` without first `git fetch origin`, should both fetch the content and set FETCH_HEAD, surely?

Reply:
The problem is when 1.18 is a new branch on remote, `git fetch origin 1.18` will fetch the commit but won't create the reference locally, so `git checkout 1.18` fails.

Reply:
Is this conversation resolved now? I am not sure I understand these intricacies of git.

Reply:
It's unfortunate to have 2 fetches, I'm sure we can find some git magic to avoid that. But IMHO it's not that bad for now. Unless @eli-schwartz has an idea?

Reply:
`git fetch origin` (without any more arguments) is always supposed to do the right thing. Why aren't we just calling that? I am not sure I understood that from the conversation.

Reply:
> The problem is when 1.18 is a new branch on remote, `git fetch origin 1.18` will fetch the commit but won't create the reference locally, so `git checkout 1.18` fails.

This works for me. I don't understand your problem but I continue to believe the commit is wrong.

Consider:

```
$ git branch -d -r origin/pythonbump
Deleted remote-tracking branch origin/pythonbump (was 754fbef3d).
$ git checkout pythonbump --
fatal: invalid reference: pythonbump
$ git fetch origin pythonbump
From https://github.com/mesonbuild/meson
 * branch                pythonbump -> FETCH_HEAD
 * [new branch]          pythonbump -> origin/pythonbump
$ git checkout pythonbump --
Previous HEAD position was b09efdbbe Revert "Do not automatically set warning flags if buildtype is 'plain'"
Branch 'pythonbump' set up to track remote branch 'pythonbump' from 'origin'.
Switched to a new branch 'pythonbump'
```

Reply:
Sorry, I think I've been too confusing. There are many cases:
- revision is a new remote branch that does not exist locally
  * `git fetch origin newbranch && git checkout newbranch` works
- revision is a branch that exists locally but got updated remotely
  * `git fetch origin newbranch && git checkout newbranch` does not work because it only checkout the local outdated branch.
  * `git fetch origin newbranch && git checkout newbranch && git reset --hard FETCH_HEAD` works.
- revision is a tag
  * `git fetch origin newtag && git checkout newtag` does not work because the tag did not get created locally.
  * `git fetch origin && git checkout newtag && git reset --hard FETCH_HEAD` does not work because FETCH_HEAD is not set.
  * `git fetch origin && git fetch origin newtag && git checkout newtag && git reset --hard FETCH_HEAD` works.

That last command actually works in all those cases without having to know if revision is a tag or branch.

Reply:
> `git fetch origin` (without any more arguments) is always supposed to do the right thing. Why aren't we just calling that? I am not sure I understood that from the conversation.

You get a problem in case 2.2 and 2.3 above: you don't know to which commit you need to update an existing outdated branch. You could do `git reset --hard origin/somebranch` but that does not work if the revision (`somebranch` here) is tag.

Reply:
> > The problem is when 1.18 is a new branch on remote, `git fetch origin 1.18` will fetch the commit but won't create the reference locally, so `git checkout 1.18` fails.
> 
> This works for me. I don't understand your problem but I continue to believe the commit is wrong.

Sorry, I meant if 1.18 is a new tag actually. Your command examples won't work in the case pythonbump is a tag:
```
$ git fetch origin pythonbump
 * tag               pythonbump -> FETCH_HEAD
$ git checkout pythonbump
error: pathspec 'pythonbump' did not match any file(s) known to git
```

Reply:
```
f() {
    git fetch origin +refs/tags/*$1:refs/tags/*$1 +refs/heads/*$1:refs/remotes/origin/*$1
}
f 0.55.3
f pythonbump
```

Does set FET_HEAD, even though it doesn't print it out.

```
case $(git cat-file -t FETCH_HEAD) in
    tag)
        ref_to_use=0.55.3
        ;;
    commit)
        ref_to_use=origin/0.55.3
        ;;
esac
```

Reply:
```
git fetch --refmap '+refs/heads/*:refs/remotes/origin/*' --refmap '+refs/tags/*:refs/tags/*' origin <ref>
```

More accurate than passing the refspec as the ref to fetch. refmap lets us add a non-default tag map for fetching.

Continue with `git cat-file -t FETCH_HEAD` to see the correct way to handle checking it out. Note tags are always going to be a detached HEAD.

Reply:
@eli-schwartz almost.... but fails in the case revision is a commit id. Forgot to document that case in my list above. The good news is I know have a unit test covering that too.

Reply:
@eli-schwartz ohh, with your updated refmap command it pass all my tests, nice! Note that `git cat-file -t FETCH_HEAD` always return 'commit' but it works without doing that, we never need `origin/` AFAIK.

Reply:
If it returns "commit" then the fetched content was a branch, if it returns "tag" then the fetched content was a tag.

I guess you could, in fact, `git reset --hard FETCH_HEAD` though. No need to figure out if 0.55.3 needs to be referred to as `origin/` if you just use `FETCH_HEAD`.

Reply:
Updated this PR with your suggestion @eli-schwartz. It works in all cases I've thought about and has the advantage of fetching a single reference instead of updating everything which could save some bandwidth. Thanks!